### PR TITLE
Make snapshot-convertor become snapshot-generator

### PR DESCRIPTION
**Pull request by alicefuzier**
### What is this PR trying to achieve?
Get rid of elasticdump and let the snapshot convertor query the index directly.
This lets us:
* avoid using the filesystem to store temporary files as elasticdump currently does
* simplify our architecture
* get rid of a whole bunch of code
* get rid of intermediate dumps in the private bucket
### Who is this change for?
⛺️ 🍝

### CODE REVIEW

File: data_api/snapshot_convertor/src/main/scala/uk/ac/wellcome/platform/snapshot_convertor/services/ConvertorService.scala [hunk 1]
```diff
@@ -81,20 +82,14 @@ class ConvertorService @Inject()(actorSystem: ActorSystem,
...
-    val toDisplayWork: ((IdentifiedWork, WorksIncludes) => DisplayWork) =
-      apiVersion match {
-        case ApiVersions.v1 => DisplayWorkV1.apply
-        case ApiVersions.v2 => DisplayWorkV2.apply
-      }
+    indexName: String,
+    toDisplayWork: (IdentifiedWork, WorksIncludes) => DisplayWork)
+    : Future[MultipartUploadResult] = {
 
     // This source generates instances of DisplayWork from the source snapshot.
```
Comment (alexwlchan) [inline data_api/snapshot_convertor/src/main/scala/uk/ac/wellcome/platform/snapshot_convertor/services/ConvertorService.scala]: nit: this comment needs updating

Comment (alicefuzier) [reply data_api/snapshot_convertor/src/main/scala/uk/ac/wellcome/platform/snapshot_convertor/services/ConvertorService.scala]: Yep, thanks, will do.

**Comment by kenoir** [comment]
Looking good so far. Great work.

**Comment by alicefuzier** [comment]
Ugh, the diff doesn't understand my renamings very well, so it's just showing everything as deleted and recreated..

File: catalogue_api/api/src/test/scala/uk/ac/wellcome/platform/api/works/ApiErrorsTest.scala [hunk 2]
```diff
@@ -235,21 +235,6 @@ class ApiErrorsTest extends ApiWorksTestBase {
     }
   }
 
-  it("returns Bad Request if you ask for a malformed identifier") {
-    ApiVersions.values.toList.foreach { version: ApiVersions.Value =>
```
Comment (alicefuzier) [inline catalogue_api/api/src/test/scala/uk/ac/wellcome/platform/api/works/ApiErrorsTest.scala]: I had to update the elastic4s version to add the streams stuff and apparently that fixed this problem as well. Now the api return 404 regardless of the characters in the identifier

Comment (alexwlchan) [reply catalogue_api/api/src/test/scala/uk/ac/wellcome/platform/api/works/ApiErrorsTest.scala]: 💃

File: data_api/snapshot_generator/src/test/scala/uk/ac/wellcome/platform/snapshot_generator/services/SnapshotServiceTest.scala [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+          result shouldBe CompletedSnapshotJob(
+            snapshotJob = snapshotJob,
+            targetLocation =
+              s"http://localhost:33333/${publicBucket.name}/$publicObjectKey"
+          )
+        }
+    }
+  }
+
+  ignore("succeeds a snapshot generation even if one of the items in the index is invalid") {
```
Comment (alicefuzier) [inline data_api/snapshot_generator/src/test/scala/uk/ac/wellcome/platform/snapshot_generator/services/SnapshotServiceTest.scala]: I actually I'm starting to think that it should crash with invalid works and this is not the behaviour that we want. Thoughts?

File: data_api/snapshot_generator/src/test/scala/uk/ac/wellcome/platform/snapshot_generator/services/SnapshotServiceTest.scala [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+            snapshotJob = snapshotJob,
+            targetLocation =
+              s"http://localhost:33333/${publicBucket.name}/$publicObjectKey"
+          )
+        }
+    }
+  }
+
+  ignore(
+    "succeeds a snapshot generation even if one of the items in the index is invalid") {
```
Comment (alicefuzier) [inline data_api/snapshot_generator/src/test/scala/uk/ac/wellcome/platform/snapshot_generator/services/SnapshotServiceTest.scala]: I actually I'm starting to think that it should crash with invalid works and this is not the behaviour that we want. Thoughts?

File: data_api/snapshot_generator/src/test/scala/uk/ac/wellcome/platform/snapshot_generator/ServerTest.scala [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+  it("shows the healthcheck message") {
+    withLocalSqsQueue { queue =>
+      withLocalSnsTopic { topic =>
+        withLocalS3Bucket { bucket =>
+          withLocalElasticsearchIndex(itemType = itemType) { indexNameV1 =>
+            withLocalElasticsearchIndex(itemType = itemType) { indexNameV2 =>
+              val flags = snsLocalFlags(topic) ++ sqsLocalFlags(queue) ++ s3LocalFlags(
+                bucket) ++ esLocalFlags(indexNameV1, indexNameV2, itemType)
+              withServer(flags) { server =>
```
Comment (kenoir) [inline data_api/snapshot_generator/src/test/scala/uk/ac/wellcome/platform/snapshot_generator/ServerTest.scala]: {{{{{{{{{{_something_}}}}}}}}}}

File: .travis.yml [hunk 6]
```diff
@@ -79,8 +79,7 @@ jobs:
 
     # Data API stack
     - env: TASK=snapshot_scheduler-test
-    - env: TASK=snapshot_convertor-test
-    - env: TASK=elasticdump-test
+    - env: TASK=snapshot_generator-test
```
Comment (alexwlchan) [inline .travis.yml]: Don’t forget to update the cached `target` directory.

Comment (alicefuzier) [reply .travis.yml]: Thanks, will do

File: data_api/snapshot_generator/src/test/scala/uk/ac/wellcome/platform/snapshot_generator/ServerTest.scala (hunk 5 above)
Comment (alicefuzier) [reply data_api/snapshot_generator/src/test/scala/uk/ac/wellcome/platform/snapshot_generator/ServerTest.scala]: 🤣 yup I should reuse the `withFixtures` wrapper that's in the FeatureTest

File: catalogue_api/api/src/test/scala/uk/ac/wellcome/platform/api/works/ApiErrorsTest.scala (hunk 2 above)
Comment (alicefuzier) [reply catalogue_api/api/src/test/scala/uk/ac/wellcome/platform/api/works/ApiErrorsTest.scala]: Actually probably deleting this test is not the correct thing to do. I should rather change it to reflect the new behaviour.

File: data_api/snapshot_generator/src/main/scala/uk/ac/wellcome/platform/snapshot_generator/source/ElasticsearchWorksSource.scala [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        logger.info(s"Received ${works.length} works from $indexName")
+        works
+      })
+      .to(Sink.ignore)
+    Source
+      .fromPublisher(
+        elasticClient.publisher(
+          search(s"$indexName/$itemType")
+            .query(termQuery("visible", true))
+            .scroll("30m")))
```
Comment (alexwlchan) [inline data_api/snapshot_generator/src/main/scala/uk/ac/wellcome/platform/snapshot_generator/source/ElasticsearchWorksSource.scala]: Name/explain this argument

File: data_api/snapshot_generator/src/test/scala/uk/ac/wellcome/platform/snapshot_generator/services/SnapshotServiceTest.scala [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+            snapshotJob = snapshotJob,
+            targetLocation =
+              s"http://localhost:33333/${publicBucket.name}/$publicObjectKey"
+          )
+        }
+    }
+  }
+
+  it(
+    "fails a snapshot generation even if one of the items in the index is invalid") {
```
Comment (alexwlchan) [inline data_api/snapshot_generator/src/test/scala/uk/ac/wellcome/platform/snapshot_generator/services/SnapshotServiceTest.scala]: "even if"?

**Pull request merged**